### ISSUE TITLE
OTP 26 -> meck issues with unloaded module

### ISSUE THREAD

**Issue by itssundeep** [erlang/otp#7479]
**Describe the bug**
In OTP 26.0.2 `meck:unload` is stuck at `code:delete(meck_util:original_name(Mod))`, we see `code_server` has a lot of messages related to `get_object_code` and some `load_module` for the same mecked module. Here the mecked module is called very often during `meck:unload`

**To Reproduce**

**Expected behavior**
In otp 25, we do not see any queued messages in code_server related to `get_object_code` or `load_module` for the same mecked module, and `code:delete` finishes instantly.

**Affected versions**
OTP 26. 

**Additional context**
- Sample messages
```
{code_call,&lt;0.64680.0&gt;,
                        {load_module,#Ref&lt;0.4109450702.1673396228.83575&gt;,
                            <some_module>,
                            "<path>/<some_module>.beam",
                            false,true}},
{code_call,&lt;0.67876.0&gt;,{get_object_code,<some_module>}},
{code_call,&lt;0.67878.0&gt;,{get_object_code,<some_module>}},
 ```

**Comment by zzydxm** [comment]
https://github.com/erlang/otp/pull/6736/files#diff-742ca058889126acdedf4d1d70c48fc534ecac2c39a1d38fc3233d487a20d158R199

Seems this change in fact add more contention in code server when unloaded module is called with high frequency.

It is because now ensure_loaded is split to 2 steps, and now code_server has to go through the PATH list many times in get_object_code until load_module is finally called.

**Comment by josevalim** [comment]
I will look into this, although I don’t understand why it could be slower. Is the call to meck:unload happening concurrently? I also assume the module is never loaded, so even in previous approaches, wouldn’t it end up with the same N calls to the code server?

**Comment by michalmuskala** [comment]
I was just looking at it this morning and I'm pretty sure I know what's the problem. The old `ensure_loaded` call would short-circuit if the module was already loaded, while `get_object_code` does not do that. So if you have a lot of concurrent calls all trying to load the module at the same time, there can be a considerable increase in the amount of work happening.

I can send a PR. I'm trying to figure out a good test case to showcase the problem.

**Comment by josevalim** [comment]
Is the goal to have something like get_obejct_code_unless_loaded? Or should we move it back to the server?

**Comment by michalmuskala** [comment]
`get_obejct_code_unless_loaded` is what I had in mind

**Comment by josevalim** [comment]
Perfect. There will still be some races but hopefully enough to address the issue. Looking forward to the PR!

**Comment by josevalim** [comment]
@michalmuskala the guaranteed way of addressing this is to do some sort of module locking. We could have a `ensure_loaded` that monitors the caller process, returns the object code, and puts all subsequent requests for said module in a queue. All further requests go into the queue. If it succeeds, we return `ok` to everyone. Otherwise, we allow the next one to try it.

**Comment by michalmuskala** [comment]
Yes, the simple approach was not enough to solve the issue in all cases - it did help a bit, but we were still seeing the issue.
I've added extra synchronisation that makes sure only one process at a time is attempting to load the module using `code:ensure_loaded`

**Comment by frazze-jobb** [comment]
Fixed with #7503
